### PR TITLE
implementation of MmapStringList

### PR DESCRIPTION
**Pull request by behrica**
Implemented for String only

### CODE REVIEW

File: test/tech/v3/datatype/mmap_list_test.clj [hunk 1]
```diff
@@ -0,0 +1,99 @@
...
+  ;; (lsize [this] 0)
+
+  (ensureCapacity [item new-size])
+
+  (addBoolean [this value])
+
+  (addDouble [this value])
+
+  (addObject [this value]
+    (let [bytes (.getBytes value)
```
Comment (behrica) [inline test/tech/v3/datatype/mmap_list_test.clj:72]: assumes String for the moment

**Comment by behrica** [comment]
My biggest doubt for the moment would be the question if the "opening every time for appendending" would be very slow.
Same for making a mmap everytime we read a  single element from it.

On the other side, I have done something similar in my original prototype for writing.
For reading I kept the "mmap object".

A primitive list could in theory allow alternate writing and reading.

**Comment by behrica** [comment]
An other key question is, if this mmaped primitive list should only work with Strings and fail on everything else.
It is probably correct to assume, that it only make sense to be used for Strings  (or other datatypes which are several bytes long for each element) Which excludes all primitive types.

It could in theory work for primitives, but it needs more storage for the positions then a primitive long buffer would use.

File: test/tech/v3/datatype/mmap_list_test.clj (hunk 1 above)
Comment (cnuernber) [reply test/tech/v3/datatype/mmap_list_test.clj:72]: totally reasonable until we get everything working.  Maybe later we just throw if it isn't a string.

File: test/tech/v3/datatype/mmap_list_test.clj [hunk 2]
```diff
@@ -0,0 +1,99 @@
...
+            )
+
+(:import
+           [tech.v3.datatype PrimitiveList Buffer]
+           )
+  )
+
+(defn append-to-file [fpath bytes]
+  (with-open [o (io/output-stream fpath :append true)]
+    (.write o bytes)))
```
Comment (cnuernber) [inline test/tech/v3/datatype/mmap_list_test.clj:16]: The output stream could be a mutable variable on the list:
`^{:unsynchronized-mutable true ^tag OutputStream} output-stream`

Same idea with the mmap handle.

Comment (behrica) [reply test/tech/v3/datatype/mmap_list_test.clj]: for output stream we would gain 20%

**Comment by behrica** [comment]
cleaned up a bit the (still simplest) working code.
 
 I will try to measure what we could gain by keeping mmap handle

**Comment by behrica** [comment]
Keeping the mmap, if possible, improves performance a lot.
I use a dirty-flag for that, which is set to true on every write.

**Comment by cnuernber** [comment]
It is odd to me that keeping the write-file open doesn't improve performance by factors, not 20%.  That doesn't jell with my experience but systems and times change.  I will be interested to mess with this soon.

Dirty flag works fine; potentially the existence of the output-stream would be the dirty flag.

**Comment by behrica** [comment]
I replaced dirt-flag by nil? -check on mmap.

I timed again the usage of an external given output-stream vs. open it every time, and see know the every time opening 2.5 times slower.

**Comment by behrica** [comment]
If you think it is in principle ok, I could start to implement https://github.com/techascent/tech.ml.dataset/issues/190 using it

**Comment by cnuernber** [comment]
I think it is definitely OK to move one from here.  Once many users start using it we will get issues then make it better but as a first cut I think it is great and it points the way to doing this generically across many datatypes.

I think once we get the tmd provably working we can merge this in an produce a new release.  That is good enough for now.

**Comment by behrica** [comment]
I needed to use FileoutputStream + sync().
The "flush" was not enough to guarantee to write to disk.
Calling it ast after each other, failed to read the file length form disk

It seems to be even faster..., half time then the BufferedOutputStream I had before

**Comment by cnuernber** [comment]
This is looking better.  The flush/sync is only necessary right before you open the mmap file, correct?  

One thing that I am slightly concerned with is that for the dataset implementation we will need to have a distinct datatype (like the Text datatype) for the datatype system to work correctly.  I wonder if we shouldn't pull the text object into dtype-next and have this object produce text objects upon read.

**Comment by behrica** [comment]
I changed to FileChannel. Seems cleaner to me, as it exposed the "position" in the file.

**Comment by cnuernber** [comment]
I would like to take this a lot further eventually so that I have generalized file-backed binary storage.  The next step is to separate the string encoding from the ability to store byte binary data.  So I would like to land this and then use it to move to a more binary-focused design.

**Comment by cnuernber** [comment]
For instance, you can now see how a file-backed primitive list could be the big byte array datastructure in the original [encoded-text](https://github.com/techascent/tech.ml.dataset/blob/b0a00c650a46888c495c12a5f12abb0ca3140f6c/src/tech/ml/dataset/text.clj#L140) design.

**Comment by behrica** [comment]
I suppose you mean to generalize the MmapString list into  a general MmapList, which takes additionally a `to-bytes-fn` and a `from bytes-fn` and can then work on any object, as long as we can define the `fns` accordingly. Correct ?

**Comment by cnuernber** [comment]
It would only work on things that are already in binary form.  So byte arrays, float arrays.  It would take a little-vs-big endianness same as native buffers.  Encoding arbitrary objects to byte form would be a layer above this such as in the the encoded text object I linked to.  But then it would just be responsible for writing binary data into a file and reading it back.

**Comment by behrica** [comment]
I generalised it via 2 function which convert from and to bytes

**Comment by behrica** [comment]
A further generalisation would now remove the direct usage of FileChannel channel and mmap. Do you want do do this step as well ?
What would we have then ?

A way to store arbitrary "blocks of bytes" in a  byte-buffer type of  backend and keep track of its positions.

Backends would differ by:
- offheap / onheap / on-disk
- unlimited / limited capacity
- extendable / non extendable

**Comment by cnuernber** [comment]
to/from bytes specializes this object to a particular class type.  The point I keep trying to make is to keep this object a generic byte store and *use* it in the encoded text object as opposed to *specialize* it here.

**Comment by cnuernber** [comment]
For instance, I would like to use this to build large arrow files.  To do that each column would first be written out to a separate file then the files combined into one large arrow file.  

Really all I need is file-backed bytes.  Anything else is something that could go in a layer above.  Adding in endian-encoding extends it to file-backed binary data but adding in a specific object encoding is something totally different and class-specific.

**Comment by behrica** [comment]
I understood.
So I went back to Filechannel dependency and made the datatype an additional parameter,

Do we need further changes at this point in time ?

**Comment by cnuernber** [comment]
That is great.  This is a good step further this direction.

**Comment by cnuernber** [comment]
Carefully thinking this over I think we should have an interface specifically for writing bytes and this object has a conversion to a read-only buffer like what you have.  That is the sort of lowest level common denominator and lots of things can be built upon that.

**Pull request merged**